rust/protocol/src/curve.rs

```diff
@@ -60,56 +85,70 @@ impl PublicKey {
+            KeyType::Curve25519 => {
                 // We allow trailing data after the public key (why?)
-                if value.len() < 32 + 1 {
-                    return Err(SignalProtocolError::BadKeyLength(KeyType::Djb, value.len()));
+                if value.len() < PUBLIC_KEY_LENGTH + 1 {
+                    return Err(SignalProtocolError::BadKeyLength(
+                        KeyType::Curve25519,
+                        value.len(),
+                    ));
                 }
-                let mut key = [0u8; 32];
-                key.copy_from_slice(&value[1..33]);
+                let mut key = [0u8; PUBLIC_KEY_LENGTH];
+                key.copy_from_slice(&value[1..(PUBLIC_KEY_LENGTH + 1)]);
                 Ok(PublicKey {
-                    key: PublicKeyData::DjbPublicKey(key),
+                    key: PublicKeyData::Curve25519PublicKey(key),
                 })
             }
         }
     }
 
+    /// Return the bytes that make up this public key.
     pub fn public_key_bytes(&self) -> Result<&[u8]> {
-        match self.key {
-            PublicKeyData::DjbPublicKey(ref v) => Ok(v),
-        }
+        Ok(self.key_data())
     }
 
+    /// Create an instance by attempting to interpret `bytes` as a [KeyType::Curve25519] public key.
     pub fn from_djb_public_key_bytes(bytes: &[u8]) -> Result<Self> {
-        match <[u8; 32]>::try_from(bytes) {
-            Err(_) => Err(SignalProtocolError::BadKeyLength(KeyType::Djb, bytes.len())),
+        match <[u8; PUBLIC_KEY_LENGTH]>::try_from(bytes) {
+            Err(_) => Err(SignalProtocolError::BadKeyLength(
+                KeyType::Curve25519,
+                bytes.len(),
+            )),
             Ok(key) => Ok(PublicKey {
-                key: PublicKeyData::DjbPublicKey(key),
+                key: PublicKeyData::Curve25519PublicKey(key),
             }),
         }
     }
 
+    /// Return a byte slice which can be deserialized with [Self::deserialize].
     pub fn serialize(&self) -> Box<[u8]> {
         let value_len = match self.key {
-            PublicKeyData::DjbPublicKey(v) => v.len(),
+            PublicKeyData::Curve25519PublicKey(v) => v.len(),
         };
-        let mut result = Vec::with_capacity(1 + value_len);
-        result.push(self.key_type().value());
+        let mut result: Vec<u8> = Vec::with_capacity(1 + value_len);
+        result.push(self.key_type().into());
         match self.key {
-            PublicKeyData::DjbPublicKey(v) => result.extend_from_slice(&v),
+            PublicKeyData::Curve25519PublicKey(v) => result.extend_from_slice(&v),
         }
         result.into_boxed_slice()
     }
 
+    /// Validate whether `signature` successfully matches `message` for this public key.
+    ///
+    /// Return `Ok(false)` if the signature fails to match or could not be read.
+    ///
+    /// *TODO: rewrite this method to consume static-sized arrays.
+    /// See <https://github.com/signalapp/libsignal-client/issues/295>.*
```

Comment:
Like the key lengths not being fixed in this API, the signature length can't really be fixed either. We've essentially got constants associated with particular key types, and pushing them into the type system would only be useful if we statically knew a key had a certain type…which we sometimes do, but often don't.